src/Surfnet/StepupSelfService/SelfServiceBundle/Controller/SamlController.php

```diff
@@ -18,29 +18,110 @@
+
+        $secondFactor     = $secondFactorService->findOneVetted($secondFactorId);
+        $secondFactorType = new SecondFactorType($secondFactor->type);
+
+        $authenticationRequest = $authenticationRequestFactory->createSecondFactorTestRequest(
+            $identity->nameId,
+            $loaResolutionService->getLoaByLevel($secondFactorType->getLevel())
+        );
+
+        $this->get('session')->set('second_factor_test_request_id', $authenticationRequest->getRequestId());
+
+        $logger = $this->get('surfnet_saml.logger')->forAuthentication($authenticationRequest->getRequestId());
+        $logger->notice('Sending authentication request to the second factor test IDP');
+
+        return $this->get('surfnet_saml.http.redirect_binding')->createRedirectResponseFor($authenticationRequest);
+    }
+
     public function consumeAssertionAction(Request $httpRequest)
     {
-        /** @var \Surfnet\SamlBundle\Http\PostBinding $postBinding */
+        $logger = $this->get('logger');
+
+        $logger->notice('Received an authentication response for testing a second factor');
+
+        $session = $this->get('session');
+
+        if (!$session->has('second_factor_test_request_id')) {
+            $logger->error(
+                'Received an authentication response for testing a second factor, but no second factor test response was expected'
+            );
+
+            throw $this->createAccessDeniedException('Did not expect an authentication response');
+        }
+
+        $initiatedRequestId = $session->get('second_factor_test_request_id');
+
+        $logger = $this->get('surfnet_saml.logger')->forAuthentication($initiatedRequestId);
+
+        $session->remove('second_factor_test_request_id');
+
         $postBinding = $this->get('surfnet_saml.http.post_binding');
 
-        /** @var \SAML2_Assertion $assertion */
-        $assertion = $postBinding->processResponse(
-            $httpRequest,
-            $this->get('surfnet_saml.remote.idp'),
-            $this->get('surfnet_saml.hosted.service_provider')
-        );
+        try {
+            $assertion = $postBinding->processResponse(
+                $httpRequest,
+                $this->get('self_service.second_factor_test_idp'),
+                $this->get('surfnet_saml.hosted.service_provider')
+            );
+
+            if (!InResponseTo::assertEquals($assertion, $initiatedRequestId)) {
+                $logger->error(
+                    sprintf(
+                        'Expected a response to the request with ID "%s", but the SAMLResponse was a response to a different request',
+                        $initiatedRequestId
+                    )
+                );
+
+                throw new AuthenticationException('Unexpected InResponseTo in SAMLResponse');
+            }
+
+            $session->getFlashBag()->add('success', 'ss.test_second_factor.verification_successful');
+        } catch (Exception $exception) {
```

Comment:
This would also catch an Exception if anything goes wrong in Gateway or client bundles (which is possible when sessions are somehow tainted during testing).